include/oneapi/dpl/experimental/kt/single_pass_scan.h

```diff
@@ -0,0 +1,436 @@
+{
+    using _Type = oneapi::dpl::__internal::__value_t<_InRange>;
+    using _FlagType = __scan_status_flag<_Type>;
+    using _FlagStorageType = typename _FlagType::_FlagStorageType;
+
+    using _KernelName = typename _KernelParam::kernel_name;
+    using _LookbackInitKernel =
+        oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<__lookback_init_kernel<_KernelName>>;
+    using _LookbackKernel =
+        oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<__lookback_kernel<_KernelName>>;
+
+    const ::std::size_t __n = __in_rng.size();
+
+    if (__n == 0)
+        return sycl::event{};
+
+    static_assert(_Inclusive, "Single-pass scan only available for inclusive scan");
+    static_assert(oneapi::dpl::unseq_backend::__has_known_identity<_BinaryOp, _Type>::value,
+                  "Only binary operators with known identity values are supported");
+
+    assert("This device does not support 64-bit atomics" &&
+           (sizeof(_Type) < 64 || __queue.get_device().has(sycl::aspect::atomic64)));
+
+    // Next power of 2 greater than or equal to __n
+    auto __n_uniform = __n;
+    if ((__n_uniform & (__n_uniform - 1)) != 0)
+        __n_uniform = oneapi::dpl::__internal::__dpl_bit_floor(__n) << 1;
+
+    // Perform a single-work group scan if the input is small
+    if (oneapi::dpl::__par_backend_hetero::__group_scan_fits_in_slm<_Type>(__queue, __n, __n_uniform))
+    {
+        return oneapi::dpl::__par_backend_hetero::__parallel_transform_scan_single_group(
+            oneapi::dpl::execution::__dpl::make_device_policy(__queue), ::std::forward<_InRange>(__in_rng),
+            ::std::forward<_OutRange>(__out_rng), __n, oneapi::dpl::__internal::__no_op{}, unseq_backend::__no_init_value<_Type>{}, __binary_op, ::std::true_type{});
+    }
+
+    constexpr ::std::size_t __workgroup_size = _KernelParam::workgroup_size;
+    constexpr ::std::size_t __data_per_workitem = _KernelParam::data_per_workitem;
+
+    // Avoid non_uniform n by padding up to a multiple of workgroup_size
+    std::uint32_t __elems_in_tile = __workgroup_size * __data_per_workitem;
+    ::std::size_t __num_wgs = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __elems_in_tile);
+
+    constexpr int __status_flag_padding = SUBGROUP_SIZE;
+    std::uint32_t __status_flags_size = __num_wgs + 1 + __status_flag_padding;
+
+    _FlagStorageType* __status_flags = nullptr;
+    _Type* __status_vals_full = nullptr;
+    _Type* __status_vals_partial = nullptr;
+
+    if constexpr (::std::is_same_v<_Type, _FlagStorageType>)
+    {
+        __status_flags = sycl::malloc_device<_FlagStorageType>(3 * __status_flags_size, __queue);
+        __status_vals_full = __status_flags + __status_flags_size;
+        __status_vals_partial = __status_flags + (2 * __status_flags_size);
+    }
+    else
+    {
+        __status_flags = sycl::malloc_device<_FlagStorageType>(__status_flags_size, __queue);
```

Comment:
If we have a different `_FlagStorageType` and `_Type`, then I think we can still use a single `malloc_device` so long as we appropriately pad it for the data-type and then reinterpret the pointer. I believe the sort-kt is currently doing this.

Do you have any objection to this?

Reply:
Good suggestion. I used some reinterpret_casts to condense this into a single allocation.